apps/studio/src/server/modules/user/inactiveUsers.service.ts

```diff
@@ -0,0 +1,198 @@
+  // we are NOT deleting the user themselves because the intention is to
+  // remove their access to interact with their sites, not remove them from Isomer
+  let sitesAndAdmins: AccountDeactivationEmailTemplateData["sitesAndAdmins"] =
+    []
+
+  try {
+    sitesAndAdmins = await db
+      .transaction()
+      .setIsolationLevel("serializable")
+      .execute(async (tx) => {
+        const deletedPermissions = await tx
+          .updateTable("ResourcePermission")
+          .where("userId", "=", user.id)
+          .where("deletedAt", "is", null)
+          .set({ deletedAt: new Date() })
+          .returningAll()
+          .execute()
+
+        // For idempotency purposes since multiple nodes may run this function at the same time
+        if (deletedPermissions.length === 0) return []
+
+        const siteIdsUserHasPermissionsFor = deletedPermissions.map(
+          (p) => p.siteId,
+        )
+
+        return await tx
+          .with("siteAdmins", (eb) =>
+            eb
+              .selectFrom("Site")
+              .leftJoin(
+                "ResourcePermission",
+                "ResourcePermission.siteId",
+                "Site.id",
+              )
+              .innerJoin("User", "User.id", "ResourcePermission.userId")
+              .where("Site.id", "in", siteIdsUserHasPermissionsFor)
+              .where("ResourcePermission.userId", "!=", user.id) // don't want to ask users to ask themselves for permissions
+              .where("ResourcePermission.userId", "not in", userIdsToDeactivate)
+              .where("ResourcePermission.deletedAt", "is", null)
+              .where("ResourcePermission.role", "=", RoleType.Admin) // should only give the admin emails to request reactivation permissions from
+              .where("User.email", "not in", ISOMER_ADMINS_AND_MIGRATORS_EMAILS) // we don't want to send emails to admins and migrators
+              .select([
+                "Site.id as siteId",
+                db.fn
+                  .agg<string[]>("array_agg", ["User.email"])
+                  .as("adminEmails"),
+              ])
+              .groupBy("Site.id"),
+          )
+          // Needed as we still want the site records even if there are no other users with permissions for that site
+          .with("baseSites", (eb) =>
+            eb
+              .selectFrom("Site")
+              .where("Site.id", "in", siteIdsUserHasPermissionsFor)
+              .select(["Site.id as siteId", "Site.name as siteName"]),
+          )
+          .selectFrom("baseSites")
+          .leftJoin("siteAdmins", "siteAdmins.siteId", "baseSites.siteId")
+          .select([
+            "baseSites.siteName",
+            sql`COALESCE("siteAdmins"."adminEmails", ARRAY[]::text[])`.as(
+              "adminEmails",
+            ),
+          ])
+          .execute()
+          .then(
+            (result) =>
+              result as AccountDeactivationEmailTemplateData["sitesAndAdmins"],
```

Comment:
[nit] avoid the `.then` and shift the cast to `sql`:

```ts
  await db.select([
            "baseSites.siteName",
            sql<
              string[]
            >`COALESCE("siteAdmins"."adminEmails", ARRAY[]::text[])`.as(
              "adminEmails",
            ),
          ])
```

main benefit is locality of information - easy to see that the cast (in code) is the same cast in sql

Reply:
good point! https://github.com/opengovsg/isomer/pull/1489/commits/559ed4d8991a71637286f68ad77b82c1faa1d657